### DISCUSSION TITLE
Code specific implementation questions?

### DISCUSSION THREAD

**Discussion by Agent6-6-6** [robbievanleeuwen/concrete-properties#32] [General]
@robbievanleeuwen

In working through what would be required to implement a NZS3101 design code option, there are a few things I'd like to raise/discuss to get some agreement on the best approach. In many cases it will probably also be relevant to implementing other similar codes which are quite different to the AS300 approach:-

- `alpha_squash` which is stored as part of the concrete material is really more specific only to a AS3600 code implementation of the interaction diagram, so I wonder if that should just be a variable existing under the AS3600 design code class? Most other codes I'm familiar with work off the max compression route cutting off the interaction diagram at the `max_comp` axial load (ACI/NZS/EC2, etc). 
    - I guess if holds no real meaning for other codes, so should it be optional if you still want it within the concrete material class so it can just be ignored for other codes?
- For the NZ concrete code, there is a different tensile strength for the concrete for an analysis for ULS vs determining deflections at the SLS. So with this in mind for doing the moment curvature you'd use the ULS direct tension strength (more of a lower characteristic value), versus if working out some properties for looking at deflections at serviceability where you'd use the modulus of rupture (more of an average characteristic value).
    - I'm not sure if there are any other codes that break it out into two different values
    - What this would practically mean is probably having two `flexural_tensile_strength` values, making the serviceability one optional and if not provided by user then is is just set to be equal to the ultimate value. Other specific functions utilising the current variable would need to be updated to refer to the ultimate or serviceability variable depending on the particular use (whether ULS or SLS analysis or parameters being calculated). 
    - Would you see that as the best approach to separate out like this? 
    - As an alternative you could create two concrete materials, but because the ULS and SLS parameters/stress strain profiles are stored under one concrete material and then one or the other would effectively be redundant, this does not seem like the best approach given the intent is one concrete material for both serviceability and ultimate analyses right?
- Do you have any aversion to feeding code specific variables as inputs directly into a design code class?
    - For example feeding in the density directly, or other code specific parameters
    - i.e. `NZS3101(density = 2300, pphr_class = 'LDPR')`?

**Comment by Agent6-6-6** [comment]
Actually on the tensile strength, I think it would work as you have it as it is only used for the cracking moment upon further investigation. For moment curvature you're inputting a specific stress/strain curve that could capture the ULS cracking behaviour.

**Reply by robbievanleeuwen** [reply]
Hi @Agent6-6-6,

Sorry for the delayed response, I've had a busy couple of weeks!

Here are my thoughts:

**1. ``alpha_squash``**

Agreed, this is an oversight from me as it should have been completely removed from the ``Concrete`` class in the ``0.3.0`` update as it no longer has any use in the main code. For AS3600, ``alpha_squash`` is calculated within the design code and is not a user supplied variable.

**2. Flexural tensile strength**

Correct, the parameter supplied to the ``Concrete`` class is used to determined cracked properties only. For moment curvature and ultimate analyses, the tensile behaviour (if any) is controlled by the respective stress-strain profile.

**3. Code specific variables**

This approach doesn't really work if you have more than one concrete strength, for example it would limit you to only one ``density`` for the entire design calculation. One of *concreteproperties*' advantages is that multiple concrete material properties can be used for a given cross-section.

What are your thoughts on creating an additional material object for code specific cases? E.g. in ``material.py`` the class ``ConcreteNZS(Concrete)`` would inherit everything from the base ``Concrete`` class and provide additional code specific parameters that you could use inside the ``NZS3101`` design code object. Inside ``NZS3101``, the ``create_concrete_material()`` would return a ``ConcreteNZS`` object rather than a ``Concrete`` object.

**Reply by Agent6-6-6** [reply]
Hi @robbievanleeuwen,

Thinking about it a little more, while a `ConcreteNZS` class works for some code specific things (variable densities being one), there are other scenarios where a code specific implementation does not really fit wholly within the material properties space.

Here I'm really thinking about in NZ code context depending on the location of the check and whether it is inside or outside a potential plastic hinge region the checks evolve in a different way (for example different maximum concrete strengths apply which would need checking, or different phi factors might apply). And those aren't really solely a concrete property thing.  

I think having another method/function within the Design Code class is potentially a better route, can define and store any additional user parameters/inputs required for that codes implementation. 

Something like a `design_code.define_code_parameters()` function that is called by the user immediately after defining the design code and offers a way of feeding additional user parameters required by a code into the design code class? 

In a NZ code sense this might define stuff like:-
- Density (though this potentially should be made part of the Concrete material class by default, after all a specific concrete mix has density and compressive strength, want to do something different then shouldn't you really be defining another concrete material altogether?)
- Type of fundamental cross section analysis required (this is potentially very NZ specific, due to the use of capacity design methods):-
    - Normal element design (i.e. normal phi, normal material properties)
    - Capacity protected element design (i.e. phi = 1.0, normal material properties)
    - Overstrength capacity design (i.e. phi = 1.0, upper characteristic material properties)
- Location of check/potential plastic hinge classification (i.e. inside or outside a potential plastic hinge region)
- possibly storing other code specific parameters I've yet to encounter....

Yet to really work this through in practice but these variables could then be specific to a particular analysis, for example could define multiple code parameter variables with different values for different analyses with the same concrete material. i.e. check a beam using Normal element design capacity against design actions, then check the overstrength capacity design based capacity for feeding overstrength actions into a column or joint design but with the same concrete/steel material modified for the upper characteristic properties.

While a user can input the three types of analysis noted above by manipulating the input materials, it kind of felt like something that should be baked in and able to selected by user to simplify the coding required at the user end.

Would something which is more general and not material specific located under the design code class make sense to store any additional parameters required for a particular analysis to that design code implementation? Then all code specific stuff is really located within the design code class code which might make maintaining the code easier as all design code specific is all in one place? 

I need to put some more thought into how this might work in practice, but interested in your thoughts, and whether this approach would be viable?

**Reply by connorferster** [reply]
I think that what you are describing is more like an application that is built on top of concreteproperties. In my mind, the design process is different from the analysis process and trying to implement so many specific design processes (e.g. the factored, nominal, probable resistances of seismic capacity design) will require specialized implementation and, essentially, it becomes an application more than a library.

My plan for implementing the Canadian code (CSS A23.3) is simply to "skin" the material classes so the user only sees CSA-style inputs and factors (familiar faces, if you will). All the capacity design stuff will be something a user has to perform on their own, whether in their own script or application. 

A simple example would be the moment magnification factor for column slenderness. The CSA implementation will provide a method to calculate the factor but the user will have to apply it to their loads separately for analysis, even though the factor itself depends on the loads. If the user wants to build a further implementation or application on top of this that just magnifies the input load for analysis automatically, then they will be able to do that easily on their own.

**Reply by robbievanleeuwen** [reply]
I think this is a healthy point of discussion - what is the scope of *concreteproperties*?

I believe *concreteproperties* should be only concerned with discrete cross-section capacity analysis, it shouldn't care/take into account what the 3D member looks like, there should be no structural analysis or member factors (as described in your example @connorferster). This is where I believe, as you suggest @connorferster, that an external package would use *concreteproperties* and extend it.

Having said that, some of the cases described by @Agent6-6-6 do pertain to cross-section capacity checks and where this is the case, I believe that *concreteproperties* should accomodate them. @Agent6-6-6 I like where you are taking your ideas for the NZ design code module, it is certainly more thorough than where I took the AS module. As a result, you will find that there are some things missing for your implementation and I am happy for you to extend the base ``DesignCode`` class as you see fit to include these parameters that you need.

I also agree that the density issue can be solved by creating a different ``Concrete`` object, however if you feel like you need a more specialised material object for the NZ module, feel free to create one!

Also keen to hear both of your thoughts on what the scope of *concreteproperties* should be!

**Reply by connorferster** [reply]
That sounds good to me. Agreed!

I am interested to see your NZ code implementation, @Agent6-6-6. It sounds like it will provide good inspiration!

**Reply by Agent6-6-6** [reply]
Hi @robbievanleeuwen, thanks for removing `alpha_squash` from the concrete material.

I've got to a point on the NZS3101 implementation where I think I do require an additional parameter stored against a material class. So hoping to do something like:-
```class SteelBarNZ(SteelBar):```

Inheriting the variable parameters from SteelBar, and adding one additional parameter via the new `SteelBarNZ` class called `phi_os`. Then call this new class when defining the steel reinforcing material using a method like the `create_steel_material()` in `AS3600` class, but modified to NZS3101 requirements.

However my python knowledge is letting me down, just cannot seem to get this to work with various `__init__()` and/or `super()` configurations. I've tried a few things I've worked through online but no success.

Are you or @connorferster able to provide a small snippit of code for this type of class to get it to inherit all the variables from `SteelBar` and add an additional phi_os parameter (which stores the material factor to elevate the yield strength to the maximum likely material strength for overstrength design). Based on what I've read this should be possible but I'm missing something I feel in getting it to work that I just can't seem to pin down.

I've tried a few things I've worked through online but no success.

How would you see this working? I was imagining this additional class being located within the specific `NZS3101` class (if that was possible) or just above it in the design_code.py file so it's all together, would that be your view of the best way to do it, or would you see it being located in the material.py file?

I kind of felt keeping it all together makes maintaining things a bit easier and doesn't clutter the material.py file with a lot of design code specific code?

**Reply by robbievanleeuwen** [reply]
Hi @Agent6-6-6, the material objects are dataclasses so you need to make sure you are using the decorators (not sure if this is the issue?) As a result they also don't need an ``__init__()`` method.

> How would you see this working?

Tricky one, but I think it should probably live in the ``material.py`` file as it is a material object. We can control how the user views this information in the documentation. This specialised material class can be included in the docs under the NZS section of the design code page, rather than appear on the materials page.

Regarding how this new class should look, here's an example. Unfortunately sphinx doesn't work very well with inherited dataclasses yet so you have to repeat all the inherited variables to get the docs to work, e.g. at the bottom of the ``material.py`` file:

```python
@dataclass
class SteelBarNZ(SteelBar):
    """Class for a steel bar material to NZS3101, treated as a lumped circular mass
    with a constant strain.

    :param name: Steel bar material name
    :param density: Steel bar density (mass per unit volume)
    :param stress_strain_profile: Steel bar stress-strain profile
    :param phi_os: TODO description here...
    :param colour: Colour of the material for rendering
    """

    name: str
    density: float
    stress_strain_profile: ssp.StressStrainProfile
    phi_os: float
    colour: str
    meshed: bool = field(default=False, init=False)
```

I haven't tested the above, but this should work. You can also have a look at how the other objects inherit from the base ``Material`` object to get an idea of how it's working. Let me know if you have any questions - happy to look over some draft code for this one!

**Reply by Agent6-6-6** [reply]
@robbievanleeuwen, yeah that `@dataclass` decorator was the issue, I was trying to inherit via and `__init__()` and `super().__init__()` route. It appeared to be all running/working but for some reason the values just were not getting stored and it was starting to drive me a little nuts! But all working now thanks! Bit of a pain to have to repeat all of the inherited variables, but so be it.

I'll probably put together a draft pull request sometime during the week if I get enough time after work next week to work on it a little more.


Couple of additional things I've noted as well which could possibly be improved upon as well:-

- For the control points on the interaction curves, it raises an exception when the number of labels and control points do not match (but only for the main curve, not the max_comp labels). 
    - It would probably be helpful in the exception to state the number of control points and labels currently specified so the user can more easily see which one is incorrect based on whatever the user is trying to ultimately achieve.
- For the control points, would it be possible to sort them into the correct order once processed rather than rely on them being entered in the correct order by the user? Would make it a little less idiot proof, especially if you're mixing and matching different types of control points I found it a bit of a pain to get it right when tinkering around with them.

**Reply by connorferster** [reply]
> > How would you see this working?
> 
> Tricky one, but I think it should probably live in the `material.py` file as it is a material object. We can control how the user views this information in the documentation. This specialised material class can be included in the docs under the NZS section of the design code page, rather than appear on the materials page.

@robbievanleeuwen @Agent6-6-6 

I do not think that this should be in `material.py` for two reasons:

1. If someone else were to do some maintenance on the NZ code implementation module, they would probably expect to see all of the files related to the code implementation in one place and not as part of the primary implementation. It would be unexpected to have to look into the primary implementation to update a code parameter.
2. I think it is good to maintain separation of scope. The additional parameter being added for the NZ material is part of a code implementation and does not affect the analysis the same way that, say, a strand material would. A strand material being so different and utilizable in different codes warrant being a new material added to `material.py`. If there was a code specific parameter required for a strand, then I would expect to see that in the code implementation.

By keeping this separation of scope, it also prevents this problem with having to create special cases in the the documentation, too. My thoughts are that any specific code implementation should be able to be its own Python package, completely independent of concreteproperties, **edit**: other than having concreteproperties as a dependency.

**Reply by robbievanleeuwen** [reply]
@connorferster a very convincing argument, I'm now in favour of placing ``SteelBarNZ`` in ``design_code.py``.

@Agent6-6-6 I'll have a look into these two issues you've raised - #39. Also wondering if you've read/have any thoughts on the discussion in #31 and how this impacts design code classes?

**Reply by Agent6-6-6** [reply]
Regarding #31, yeah I agree with the requirement to show the M/N diagrams at a constant load angle if that's what you're specifically referring to, I was just seeing how the conversation would pan out. 

A constant load angle is really what I'd expect to be returned for design purposes. I don't think it really will have any bearing on specific design code classes, as you're just calling the methods to display the M/N or Mx/My interaction diagrams from a design code. But generally a true vertical slice through the 3D surface would be considered the normal way of displaying and interpreting the results, pretty sure commercial packages do this (SPColumn for example).

>a very convincing argument, I'm now in favour of placing SteelBarNZ in design_code.py.

For now I've put it as a subclass within NZS3101 class, i.e. `NZS3101.SteelBarNZ()`, my thinking is then it stays with the particular design code class, and ultimately I'd imagine you might want to break any implemented codes into their own file and leave the base `DesignCode` class in it's own file, throw all the actual design codes into a subfolder. In that way it's about as separate as you can make it without breaking it out into a different package (which I'm not really in favour of from a usability/maintenance perspective).

**Comment by connorferster** [comment]
Agreed, I don’t think it makes sense to have it actually be a separate package. I like the idea that, when you install concreteproperties, it comes with some design codes!
